warp10/src/main/java/io/warp10/script/functions/FETCH.java

```diff
@@ -235,32 +236,83 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
+
+      Metadata tmeta = new Metadata();
+      tmeta.setName("");
+      tmeta.setLabels(tokenSelectors);
+      
+      // Build a selector matching all classes
+      String tselector = "~.*" + GTSHelper.buildSelector(tmeta, true);
+      MetadataSelectorMatcher matcher = new MetadataSelectorMatcher(tselector);
+      
+      //
+      // Build a selector
       for (Metadata m: metas) {
         if (null == m.getLabels()) {
           m.setLabels(new HashMap<String,String>());
         }
-        m.getLabels().remove(Constants.PRODUCER_LABEL);
-        m.getLabels().remove(Constants.OWNER_LABEL);
-        m.getLabels().remove(Constants.APPLICATION_LABEL);
-        m.getLabels().putAll(Tokens.labelSelectorsFromReadToken(rtoken));
-                
-        if (m.getLabels().containsKey(Constants.PRODUCER_LABEL) && '=' == m.getLabels().get(Constants.PRODUCER_LABEL).charAt(0)) {
-          m.getLabels().put(Constants.PRODUCER_LABEL, m.getLabels().get(Constants.PRODUCER_LABEL).substring(1));
-        } else if (m.getLabels().containsKey(Constants.PRODUCER_LABEL)) {
-          throw new WarpScriptException(getName() + " provided token is incompatible with '" + PARAM_GTS + "' parameter, expecting a single producer.");
-        }
         
-        if (m.getLabels().containsKey(Constants.OWNER_LABEL) && '=' == m.getLabels().get(Constants.OWNER_LABEL).charAt(0)) {
-          m.getLabels().put(Constants.OWNER_LABEL, m.getLabels().get(Constants.OWNER_LABEL).substring(1));
-        } else {
-          throw new WarpScriptException(getName() + " provided token is incompatible with '" + PARAM_GTS + "' parameter, expecting a single owner.");
+        //
+        // If the Metadata have producer/owner/app labels, check if 'matcher' would select them
+        //
+        
+        boolean matches = false;
+        
+        if (m.getLabels().containsKey(Constants.PRODUCER_LABEL)
+            && m.getLabels().containsKey(Constants.OWNER_LABEL)
+            && m.getLabels().containsKey(Constants.APPLICATION_LABEL)) {
+          matches = matcher.matches(m);
         }
         
-        if (m.getLabels().containsKey(Constants.APPLICATION_LABEL) && '=' == m.getLabels().get(Constants.APPLICATION_LABEL).charAt(0)) {
-          m.getLabels().put(Constants.APPLICATION_LABEL, m.getLabels().get(Constants.APPLICATION_LABEL).substring(1));
-        } else {
-          throw new WarpScriptException(getName() + " provided token is incompatible with '" + PARAM_GTS + "' parameter, expecting a single application.");
+        //
+        // If the metadata would not get selected by the provided token
+        // force the producer/owner/app to be that of the token
+        //
+        
+        if (!matches) {
+          m.getLabels().remove(Constants.PRODUCER_LABEL);
+          m.getLabels().remove(Constants.OWNER_LABEL);
+          m.getLabels().remove(Constants.APPLICATION_LABEL);
+          m.getLabels().putAll(tokenSelectors);
```

Comment:
You can't use that because it will add a `=` before each label value, which won't be overridden in case there is a producer and no owner or no producer and an owner.
At the end the labelID will be wrong.